src/python/modules/Tensorflow/gmm_objective.py

```diff
@@ -0,0 +1,98 @@
+import math
+from scipy import special
+import tensorflow as tf
+
+from modules.Tensorflow.utils import shape
+
+
+
+def logsumexp(x):
+    mx = tf.reduce_max(x)
+    emx = tf.exp(x - mx)
+    return tf.math.log(tf.reduce_sum(emx)) + mx
+
+
+
+def log_gamma_distrib(a, p):
+    return special.multigammaln(a, p)
+
+
+
+def sqsum(x):
+    return tf.reduce_sum(x ** 2)
+
+
+
+def log_wishart_prior(p, wishart_gamma, wishart_m, sum_qs, Qdiags, icf):
+    n = p + wishart_m + 1
+    k = shape(icf)[0]
+    out = 0
+    for ik in range(k):
+        frobenius = sqsum(Qdiags[ik, :]) + sqsum(icf[ik, p:])
+        out += 0.5 * wishart_gamma * wishart_gamma * frobenius - \
+               wishart_m * sum_qs[ik]
+
+    log2 = tf.math.log(tf.constant(2.0, dtype = tf.float64))
+    C = n * p * (tf.math.log(wishart_gamma) - 0.5 * log2) - \
+        log_gamma_distrib(0.5 * n, p)
+
+    return out - k * C
+
+
+
+def constructL(d, icf):
+    constructL.Lparamidx = d
+
+    def make_L_col(i):
+        nelems = d - i - 1
+        col = tf.concat([
+            tf.zeros(i + 1, dtype=tf.float64),
+            icf[constructL.Lparamidx:(constructL.Lparamidx + nelems)]
+        ], 0)
+
+        constructL.Lparamidx += nelems
+        return col
+
+    columns = [ make_L_col(i) for i in range(d) ]
+    return tf.stack(columns, -1)
+
+
+
+def Qtimesx(Qdiag, L, x):
+    res = Qdiag * x
+
+    for i in range(shape(L)[0]):
```

Comment:
Would this be better as a matrix multiply?

Reply:
As I understand it's not a matrix multiplication exactly. It's a componentwise sum of componentwise multiplication x by L columns, but I'll try to find a way to do it more "tensorflowly"

Reply:
Changed